File: docs/build-log/issue-639-section-group-concurrency/mutations.mjs
import assert from 'node:assert/strict';
import { readFileSync, writeFileSync } from 'node:fs';
import { spawnSync } from 'node:child_process';
const dir = 'docs/build-log/issue-639-section-group-concurrency';
const store = 'mcp/src/clinical-graph/finding-section-group-store.ts';
const endpoint = 'mcp/src/clinical-graph/finding-section-group-endpoint.ts';
const tests = 'mcp/tests/findingSectionGroupConcurrency.test.ts';
const oldTests = 'mcp/tests/findingSectionGroup.test.ts';
const catalogueHeader = '"If-None-Exist": `identifier=${encodeURIComponent(`${FINDING_SECTION_GROUP_IDENTIFIER_SYSTEM}|${validated.groupKey}`)}`';
const settings = 'ui/src/components/settings/FindingSectionGroupsSettings.tsx';
const uiTests = 'ui/tests/findingSectionGroups.test.tsx';
const cases = [
 ['UI-new-edit-error', settings, 'function edit(group: FindingSectionGroup) {\n    setError(null);', 'function edit(group: FindingSectionGroup) {', uiTests, 'G1 Settings deactivate'],
 ['UI-new-create-error', settings, 'function create() {\n    setError(null);', 'function create() {', uiTests, 'G1 Settings edit'],
 ['UI-create-version', settings, '              expectedVersion: null,\n', '', uiTests, 'S1b G6 settings'],
 ['UI-edit-version', settings, 'expectedVersion: editing.original?.versionId ?? null', 'expectedVersion: null', uiTests, 'G1 Settings edit'],
 ['UI-active-version', settings, 'expectedVersion: group.versionId', 'expectedVersion: null', uiTests, 'G1 Settings deactivate'],
 ['UI-conflict-message', settings, 'throw clinicalGraphResponseError(response, result, `Section-group save failed: ${response.status}`)', 'throw new Error(result.error)', uiTests, 'G1 Settings edit'],
 ['G1', store, 'const writeVersion = expectedVersion;', 'const writeVersion = existing?.meta?.versionId ?? null;', tests, 'G1'],
 ['G2', store, catalogueHeader, '"X-Disabled-Condition": "fault-injection"', tests, 'G2'],
 ['G3', store, catalogueHeader, '"X-Disabled-Condition": "fault-injection"', tests, 'G3'],
 ['G4', store, 'if (!json || (JSON.parse(json) as { writeToken?: string }).writeToken !== writeToken) throw new FindingSectionGroupConcurrentEditError();', '', tests, 'G2'],
 ['G5', store, '"If-None-Exist": new URLSearchParams({', '"X-Disabled-Condition": new URLSearchParams({', tests, 'G5'],
 ['G6', store, 'if ((error as { status?: number }).status === 412) throw new FindingSectionGroupConcurrentEditError();', 'if ((error as { status?: number }).status === 412) return { ...validated, versionId: expectedVersion };', oldTests, 'stale group and encounter-override writes'],
 ['G7', endpoint, 'if (sectionKeys.length) return', 'if (false) return', oldTests, 'S1 G1 removal'],
];
function run(file, pattern) {
 const args = ['--import','tsx','--test',`--test-name-pattern=${pattern}`,file];
 const result = spawnSync(process.execPath,args,{encoding:'utf8'});
 const out = result.stdout + result.stderr;
 return { command: `node ${args.join(' ')}`, exit: result.status, summary: out.split('\n').filter(line => /^(# (tests|pass|fail|skipped)|not ok |ok )/.test(line)) };
}
const results = [];
for (const [guard, file, from, to, test, pattern] of cases) {
 const original = readFileSync(file,'utf8');
 assert.equal(original.split(from).length,2,`${guard} unique mutation anchor`);
 let red;
 try { writeFileSync(file,original.replace(from,to)); red=run(test,pattern); assert.notEqual(red.exit,0,`${guard} must fail`); }
 finally { writeFileSync(file,original); }
 const green=run(test,pattern); assert.equal(green.exit,0,`${guard} restored must pass`);
 results.push({guard,red,green}); console.log(`${guard}: red ${red.exit}; restored green ${green.exit}`);
}
writeFileSync(`${dir}/mutations.json`,JSON.stringify(results,null,2)+'\n');

File: docs/build-log/followup-s3b1-shape-record/proof/mutations.py
from pathlib import Path
import subprocess, json, re, shlex
root=Path(__file__).resolve().parents[4]
store='mcp/src/clinical-graph/exam-scope-store.ts'
endpoint='mcp/src/clinical-graph/exam-overview-endpoint.ts'
board='ui/src/components/charting/ExamOverviewBoard.tsx'
projection='mcp/src/clinical-graph/exam-overview-projection.ts'
out=root/'docs/build-log/followup-s3b1-shape-record/mutations.json'
server='mcp/tests/examShapeRecord.test.ts'; scopes='mcp/tests/examScopeStore.test.ts'; ui='ui/tests/examOverviewBoard.test.tsx'; ends='mcp/tests/examOverviewEndpoint.test.ts'
write_existing=('return this.write(encounterId, "comprehensive", setBy, null, undefined, shape);','return this.write(encounterId, existing ? parseScope(existing).examScope as ExamScope : "comprehensive", setBy, existing?.meta?.versionId ?? null, existing, shape);')
cases=[
 ('G1',store,[('if (existing) return parseScope(existing);',''),write_existing],'S3b1 G1',[server]),
 ('G2a',store,[('const shape = ["profilesApplied", "sectionsOpen", "shapedAt"].some(key => key in value) ? shapeSchema.parse(value) : undefined;','const shape = shapeSchema.parse(value);')],'S3b1 G2',[ui]),
 ('G2b',store,[('if (existing) return parseScope(existing);','if (existing && parseScope(existing).shapedAt) return parseScope(existing);'),write_existing],'S3b1 G2',[ui]),
 ('G3',board,[('const editor = editorForFinding(group, editorEntries);', 'const editor = editorForFinding(group, editorEntries);\n    if (!(projection.sectionsOpen ?? []).includes(editor?.id ?? \"\")) continue;'),('return opened.has(editor.id) || groups.some','return (projection.sectionsOpen ?? []).includes(editor.id) && (opened.has(editor.id) || groups.some'),('(scopeDrawsRow && !editor.readOnly && (!definition.singleBlank || index === 0));','(scopeDrawsRow && !editor.readOnly && (!definition.singleBlank || index === 0)));')],'S3b1 G3',[ui]),
 ('G4-stale',store,[('if ((existing?.meta?.versionId ?? null) !== expectedVersion) throw concurrentEdit();','expectedVersion = existing?.meta?.versionId ?? null;')],'S2a scope store guards',[scopes]),
 ('G4-create',store,[('this.fhir.create(resource, { "If-None-Exist": `identifier=${encodeURIComponent(`${SCOPE_SYSTEM}|${encounterId}`)}` })','this.fhir.create(resource)')],'S3b1 G4 conditional',[server]),
 ('G4-token',store,[('if (persistedValue.writeToken !== writeToken) throw concurrentEdit();','')],'S3b1 G4 write-token',[server]),
 ('G5',store,[('({ profileKey, version, versionId })),','({ profileKey, versionId })),')],'S3b1 G5',[server]),
 ('G6',projection,[('const policy = Object.hasOwn(registry, examScope)','const visitTypeCategoryId = "medical";\n  const policy = Object.hasOwn(registry, visitTypeCategoryId)'),('? registry[examScope]','? registry[visitTypeCategoryId]')],'S2a G7',[ends]),
 ('G8',endpoint,[('// Opening the board must not depend on bookkeeping persistence.','throw error;\n        // Opening the board must not depend on bookkeeping persistence.')],'S3b1 G8',[ends,ui]),
 ('conflict-reread',endpoint,[('scope = await scopeStore.get(encounterId); } catch','scope = { examScope: "comprehensive" }; } catch')],'S3b1 conflict',[ends]),
 ('diagnosis-match',endpoint,[('families.has(normalizeFamily(family))','false')],'S3b1 overview freezes',[ends]),
]
results=[]
for name,file,replacements,pattern,tests in cases:
 path=root/file; original=path.read_text(); mutated=original
 for old,new in replacements:
  assert old in mutated,(name,old)
  mutated=mutated.replace(old,new,1)
 cmd=['node','--import','tsx','--test','--test-name-pattern='+pattern,*tests]
 def run():
  result=subprocess.run(cmd,cwd=root,capture_output=True,text=True)
  text=result.stdout+result.stderr
  summary='\n'.join(line for line in text.splitlines() if re.match(r'^(# (tests|pass|fail|skipped)|not ok)',line))
  return result.returncode,summary
 try:
  path.write_text(mutated); red_code,red=run()
 finally: path.write_text(original)
 green_code,green=run()
 results.append({'guard':name,'command':shlex.join(cmd),'redExit':red_code,'red':red,'greenExit':green_code,'green':green})
 out.write_text(json.dumps(results,indent=2)+'\n')
 print(name, 'RED',red_code,'GREEN',green_code,flush=True)
 assert red_code!=0 and green_code==0,(name,red,green)
